lib/CL/pocl_llvm_dynlib.cc

```diff
@@ -65,34 +66,87 @@ void *pocl_dynlib_symbol_address(void *, const char *SymbolName) {
-                         (LPCWSTR)&Address, &Hm) == 0) {
+                         (LPCWSTR)Address, &Hm) == 0) {
     int Ret = GetLastError();
-    POCL_MSG_WARN("GetModuleHandleEx failed, error = %d ; trying fallback\n", Ret);
+    POCL_MSG_WARN("GetModuleHandleEx failed, error = %d; trying fallback\n",
+                  Ret);
 
-    // undocumented hack from https://stackoverflow.com/a/2396380
+    // Undocumented hack from https://stackoverflow.com/a/2396380
     MEMORY_BASIC_INFORMATION mbi;
-    size_t Len = VirtualQuery(Address, &mbi, sizeof(mbi));
-    if (Len != sizeof(mbi)) {
-      POCL_MSG_ERR("VirtualQuery failed", Ret);
+    if (VirtualQuery(Address, &mbi, sizeof(mbi)) == 0) {
+      Ret = GetLastError();
+      POCL_MSG_ERR("VirtualQuery fallback failed, error = %d\n", Ret);
       return nullptr;
     }
     Hm = (HMODULE)mbi.AllocationBase;
   }
 
-  WCHAR wpath[MAX_PATH]{};
+  // Get the path of the module
+  WCHAR wpath[MAX_PATH];
   if (GetModuleFileNameW(Hm, wpath, ARRAYSIZE(wpath)) == 0) {
     int Ret = GetLastError();
     POCL_MSG_ERR("GetModuleFileName failed, error = %d\n", Ret);
     return nullptr;
   }
 
-  static char path[MAX_PATH];
-  WideCharToMultiByte(CP_UTF8, 0, wpath, -1, path, ARRAYSIZE(path), NULL, NULL);
-  POCL_MSG_PRINT_INFO("pocl_dynlib_symbol_adress: using DLL path: %s \n", path);
-  return path;
+  // Open a the file to get a handle
+  HANDLE hFile =
+      CreateFileW(wpath, GENERIC_READ,
+                  FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+                  OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+  if (hFile == INVALID_HANDLE_VALUE) {
+    int Ret = GetLastError();
+    POCL_MSG_ERR("CreateFile failed, error = %d\n", Ret);
+    return nullptr;
+  }
+
+  // Get the final path, resolving any symlinks
+  std::vector<WCHAR> wfinal_buf(MAX_PATH);
+  DWORD path_len = GetFinalPathNameByHandleW(
+      hFile, wfinal_buf.data(), wfinal_buf.size(), VOLUME_NAME_DOS);
+  if (path_len > wfinal_buf.size()) {
+    wfinal_buf.resize(path_len);
+    path_len = GetFinalPathNameByHandleW(hFile, wfinal_buf.data(),
+                                         wfinal_buf.size(), VOLUME_NAME_DOS);
+  }
+  CloseHandle(hFile);
+  if (path_len == 0) {
+    int Ret = GetLastError();
+    POCL_MSG_ERR("GetFinalPathNameByHandle failed, error = %d\n", Ret);
+    return nullptr;
+  }
+
+  // Get rid of the long path prefix
+  std::wstring wfinal(wfinal_buf.data(), path_len);
+  if (wfinal.rfind(L"\\\\?\\", 0) == 0) {
+    wfinal = wfinal.substr(4);
+  }
+
+  // Convert the wide string to a UTF-8 multi-byte string
+  thread_local std::string final;
+  int required_size = WideCharToMultiByte(CP_UTF8, 0, wfinal.c_str(), -1,
+                                          NULL, 0, NULL, NULL);
+  if (required_size == 0) {
+    int Ret = GetLastError();
+    POCL_MSG_ERR("WideCharToMultiByte (size check) failed, error = %d\n", Ret);
+    return nullptr;
+  }
+  final.resize(required_size);
```

Comment:
The `required_size` holds number of bytes, right? The `wstring::resize()` takes number of characters so this probably result in an oversized string.

Reply:
`final` is an `std::string`, not a `wstring`, so resizing with the number of bytes should match the number of chars, no?

Reply:
Right, I got this mixed up with `wfinal` (again).